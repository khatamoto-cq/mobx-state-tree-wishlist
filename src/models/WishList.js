import {types} from "mobx-state-tree"

const data = {
  "name": "Chronicles of Narnia Box Set - C.S. Lewis",
  "price": 28.73,
  "image": "https://images-na.ssl-images-amazon.com/images/I/51a7xaMpneL._SX329_BO1,204,203,200_.jpg"
}

export const WishListItem = types
  .model({
    name: types.string,
    price: types.number,
    image: ""
  })
  .actions(self => ({
    changeName(newName) {
      self.name = newName
    },
    changePrice(newPrice) {
      self.price = newPrice
    },
    changeImage(newImage) {
      self.image = newImage
    },
  }))

export const WishList = types
  .model({
    items: types.optional(types.array(WishListItem), [])
  })
  .actions(self => ({
    add(item) {
      self.items.push(item)
    }
  }))
  .views(self => ({
    get totalPrice() {
      return self.items.reduce((sum, entry) => sum + entry.price, 0)
    }
  }))
