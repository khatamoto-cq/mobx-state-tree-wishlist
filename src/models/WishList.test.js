import {getSnapshot, onSnapshot, onPatch} from "mobx-state-tree";
import {WishListItem, WishList} from "./WishList";

it("can create a instance of a model", () => {
  const item = WishListItem.create({
    "name": "Chronicles of Narnia Box Set - C.S. Lewis",
    "price": 28.73,
  })

  expect(item.price).toBe(28.73)
  expect(item.image).toBe("")
  item.changeName("Narnia")
  expect(item.name).toBe("Narnia")
})

it("can create a wishlist", () => {
  const list = WishList.create({
    items: [
      {
        "name": "Chronicles of Narnia Box Set - C.S. Lewis",
        "price": 28.73
      }
    ]
  })

  expect(list.items.length).toBe(1)
  expect(list.items[0].price).toBe(28.73)
})

it("can add new items", () => {
  const list = WishList.create()
  const patches = []
  onSnapshot(list, patche => {
    patches.push(patche)
  })

  list.add(WishListItem.create({
    name: "Chesterton",
    price: 10
  }))

  list.items[0].changeName("Book of G.K Chesterton")

  expect(patches).toMatchSnapshot()
})
