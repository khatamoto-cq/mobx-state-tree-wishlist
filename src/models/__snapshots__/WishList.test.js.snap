// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add new items 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "Chesterton",
        "price": 10,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "Book of G.K Chesterton",
        "price": 10,
      },
    ],
  },
]
`;
